bftengine/include/bftengine/CryptoManager.hpp

```diff
@@ -30,73 +33,134 @@ class CryptoManager : public IKeyExchanger, public IMultiSigKeyGenerator {
+    auto sys = create(sn);
     // the +1 is due to Crypto system starts counting from 1
-    multiSigCryptoSystem_->updateVerificationKey(verificationKey, signerIndex + 1);
-    init();
+    sys->sys_->updateVerificationKey(verificationKey, signerIndex + 1);
+    sys->init();
+  }
+
+  void onCheckpoint(const uint64_t& checkpoint) {
+    LOG_INFO(logger(), "chckp: " << checkpoint);
+    clearOldKeys();
   }
 
  private:
-  CryptoManager(Cryptosystem* cryptoSys) : multiSigCryptoSystem_{cryptoSys} { init(); }
+  /**
+   *  Holds Cryptosystem, signers and verifiers per checkpoint
+   */
+  struct CryptoSys {
+    CryptoSys(Cryptosystem* cs) : sys_(cs) {}
+    CryptoSys(const CryptoSys&) = delete;
+    std::unique_ptr<Cryptosystem> sys_;
+    std::shared_ptr<IThresholdSigner> thresholdSigner_;
+    // signer and verifier of a threshold signature (for threshold N-fVal-cVal out of N)
+    std::shared_ptr<IThresholdVerifier> thresholdVerifierForSlowPathCommit_;
+
+    // verifier of a threshold signature (for threshold N-cVal out of N)
+    // If cVal==0, then should be nullptr
+    std::shared_ptr<IThresholdVerifier> thresholdVerifierForCommit_;
+
+    // verifier of a threshold signature (for threshold N out of N)
+    std::shared_ptr<IThresholdVerifier> thresholdVerifierForOptimisticCommit_;
+
+    void init() {
+      std::uint16_t f{ReplicaConfig::instance().getfVal()};
+      std::uint16_t c{ReplicaConfig::instance().getcVal()};
+      std::uint16_t numSigners{ReplicaConfig::instance().getnumReplicas()};
+      thresholdSigner_.reset(sys_->createThresholdSigner());
+      thresholdVerifierForSlowPathCommit_.reset(sys_->createThresholdVerifier(f * 2 + c + 1));
+      thresholdVerifierForCommit_.reset(sys_->createThresholdVerifier(f * 3 + c + 1));
+      thresholdVerifierForOptimisticCommit_.reset(sys_->createThresholdVerifier(numSigners));
+    }
+  };
+
+  // accessing existing Cryptosystems
+  std::shared_ptr<CryptoSys> get(const SeqNum& sn) const {
+    // find last chckp that is less than a chckp of a given sn
+    uint64_t chckp = (sn - 1) / checkpointWindowSize;
+    auto it = cryptoSystems_.rbegin();
+    while (it != cryptoSystems_.rend()) {
+      if (it->first <= chckp) {
+        LOG_INFO(logging::getLogger("concord.bft.crypto-mgr"), KVLOG(sn, chckp, it->first, it->second));
+        return it->second;
+      }
+      it++;
+    }
+    ConcordAssert(false && "should never reach here");
+  }
+  // create CryptoSys for sn if still doesn't exist
+  std::shared_ptr<CryptoSys> create(const SeqNum& sn) {
+    // Cryptosystem for this sn will be activated upon reaching a second checkpoint from now
+    uint64_t chckp = sn / checkpointWindowSize + 2;
+    if (auto it = cryptoSystems_.find(chckp); it != cryptoSystems_.end()) return it->second;
+    // copy construct new Cryptosystem from a last one as we want it to include all the existing keys
+    Cryptosystem* cs = new Cryptosystem(*cryptoSystems_.rbegin()->second->sys_.get());
```

Comment:
nit: why not creating ```std::make_shared<CryptoSys>(new Cryptosystem(*cryptoSystems_.rbegin()->second->sys_.get()))```

Reply:
I pass raw pointer to a CryptoSustemWrapper(), which owns it by creating a sp